core/src/main/java/org/ehcache/events/CacheEventNotificationServiceImpl.java

```diff
@@ -122,48 +147,92 @@ public void setStoreListenerSource(Cache<K, V> source) {
   }
 
   @Override
-  public void onEvent(final CacheEvent<K, V> event) {
+  public void onEvent(CacheEvent<K, V> event) {
     final EventType type = event.getType();
     LOGGER.trace("Cache Event notified for event type {}", type);
-    Map<EventListenerWrapper, Future<?>> notificationResults = 
-        new HashMap<CacheEventNotificationServiceImpl.EventListenerWrapper, Future<?>>(registeredListeners.size());
-    
-    for (final EventListenerWrapper wrapper: registeredListeners) {
-      if (!wrapper.config.fireOn().contains(type)) {
-        continue;
-      }
-      Runnable notificationTask = new Runnable() {
-        @Override
-        public void run() {
-          CacheEventListener<K, V> listener = wrapper.getListener();
-          listener.onEvent(event);
+    CacheEventWrapper<K, V> cacheEventWrapper = new CacheEventWrapper<K, V>(event);
+    eventThreadLocal.get().add(cacheEventWrapper);
+    markFailedCacheEventsInThreadLocal();
+    if (!orderedListenersSet.isEmpty()) {
+      ConcurrentLinkedQueue<CacheEventWrapper> eventWrapperQueue = getEventQueue(event.getKey());
+      eventWrapperQueue.add(cacheEventWrapper);
+    }
+  }
+  
+  private void dispatchEvent(CacheEventWrapper cacheEventWrapper) {
+    cacheEventWrapper.markFireable();
+    ConcurrentLinkedQueue<CacheEventWrapper> eventWrapperQueue = getEventQueue((K)cacheEventWrapper.cacheEvent.getKey());
+    orderedEventWait(cacheEventWrapper, eventWrapperQueue);
+    if (!orderedListenersSet.isEmpty()) {
```

Comment:
Already hinted at this elsewhere, but what happens when that value changes in the middle of processing? when it becomes true? when it becomes false?